lib/IRGen/IRGenModule.cpp

```diff
@@ -916,9 +918,9 @@ void IRGenModule::emitAutolinkInfo() {
                                        }),
                         AutolinkEntries.end());
 
-  if (TargetInfo.OutputObjectFormat == llvm::Triple::COFF ||
-      TargetInfo.OutputObjectFormat == llvm::Triple::MachO ||
-      Triple.isPS4()) {
+  if ((TargetInfo.OutputObjectFormat == llvm::Triple::COFF &&
+       (!Triple.isOSCygMing() || Triple.getArch() == llvm::Triple::thumb)) ||
```

Comment:
Test script - https://github.com/apple/swift/blob/master/test/IRGen/autolink-coff.swift - asserts the target `thumbv7--windows-gnu` to use the MSVC-style autolink, which uses `.drectve` section.
I created the `autolink-coff-x86.swift` for MinGW (`x86--windows-gnu`) to test the Linux-style autolink.
If we drop/modify the autolink test case for `thumbv7--windows-gnu` , the code will be simpler.